crates/subspace-fraud-proof/src/invalid_state_transition_proof.rs

```diff
@@ -235,79 +247,136 @@ impl<Client, Block> PreStateRootVerifier<Client, Block> {
     }
 }
 
-impl<Client, Block> VerifyPreStateRoot for PreStateRootVerifier<Client, Block>
+impl<Client, Block> VerifyPrePostStateRoot for PrePostStateRootVerifier<Client, Block>
 where
     Block: BlockT,
     Client: ProvideRuntimeApi<Block> + HeaderBackend<Block>,
     Client::Api: ReceiptsApi<Block, domain_runtime_primitives::Hash>,
 {
     fn verify_pre_state_root(
         &self,
-        domain_id: DomainId,
-        receipt_hash: H256,
-        pre_state_root: H256,
-        _execution_phase: &ExecutionPhase,
-    ) -> bool {
-        let trace = self
-            .client
-            .runtime_api()
-            .execution_trace(self.client.info().best_hash, domain_id, receipt_hash)
-            .unwrap_or_default();
+        invalid_state_transition_proof: &InvalidStateTransitionProof,
+    ) -> Result<(), VerificationError> {
+        let InvalidStateTransitionProof {
+            domain_id,
+            bad_receipt_hash,
+            parent_number,
+            parent_hash,
+            pre_state_root,
+            execution_phase,
+            ..
+        } = invalid_state_transition_proof;
 
-        // TODO: Get trace index from execution phase
-        let trace_index_of_pre_state_root = 0usize;
+        let trace = self.client.runtime_api().execution_trace(
+            self.client.info().best_hash,
+            *domain_id,
+            *bad_receipt_hash,
+        )?;
```

Comment:
How about moving this to the `ExecutionPhase::ApplyExtrinsic(...)| ExecutionPhase::FinalizeBlock` branch since the execution trace is unneeded for `ExecutionPhase::InitializeBlock`, non-blocker.